Add message extras docs (#22)
gotify/website: docs/msgextras.md

---
id: msgextras
title: Message Extras
---

> Extras in messages can be used to extend clients. This page lists properties supported by the official gotify/\* clients.

Reviewer: ```suggestion
Extras are JSON structured data that is carried together with the message and forwarded to all clients. They are then parsed by the clients and changes their behavior. There are some extra properties which are used by the official gotify/\* clients, while others are open for end-user customization.
```

`extend clients` sounds a little ambiguous.

Extras are stored in a key-value scheme and are only accepted in `POST /message` requests
with `application/json` content-type. Keys should be in the following format: `<top-namespace>::[<sub-namespace>::]<action>`

> Here's an example how the message with extras looks like. More information can be found in the [swagger-docs](../api-docs).

Reviewer: ```suggestion
Here's an example of how the message with extras looks like. For example, this message could be received by a home automation server, which decodes the extras fields as defined and perform automated tasks as defined by the user.
```

I think the swagger docs did not say anything more than this documentation. No need to refer it to the user again.

I also added a short description of how the extras of this example might works.

```json
{
  "title": "my title",

>   "message": "my message",

Reviewer: ```suggestion
  "message": "Welcome home!",
```
Looks nicer and more in accordance with the context of this message:)

  "extras": {
    "home::appliances::thermostat::change_temperature": { "temperature": 23 },
    "home::appliances::lighting::on": { "brightness": 15 }
  }
}
```

> ## Message ContentType

Reviewer: ```suggestion
## Message Rendering
```

In the future, there might be more than `contentType` under `client::display` namespace that specifies other perspectives of the display of messages.

This setting changes how clients render the content of a message.

### Example

```json
{
  "extras": {
    "client::display": {
      "contentType": "text/plain"
    }
  }
}
```

> ### Values

Reviewer: I think this part can be organized in this way, I think it looks cleaner when we added more extra fields:

## Extras
Extras are used to carry extra information, change how clients behave, etc. 
Extras are stored in a key-value scheme and are only accepted in `POST /message` requests with `application/json` content-type. 

### Namespaces

Keys under `.extras` should be in the following format: `<top-namespace>::[<sub-namespace>::]<action>`. Some of the namespaces are used by official clients:

|   Namespace| Description |
|                ---|            ---|
|         [`client::display`](#client::display) | Changes how client dispays information|
|         `client::*` | Reserved|
|         `android::*`| Reserved|
|         `ios::*`| Reserved|
|        `server::*`| Reserved|
|         All Other|  Defined by end-users|

#### `client::display`

##### Definition
| contentType     | description                                                                                         |
| --------------- | --------------------------------------------------------------------------------------------------- |
| `text/plain`    | **Default**; The message will be rendered as plain text. Links may be highlighted and be clickable. |
| `text/markdown` | The message should be rendered as markdown. **HTML should be ignored**.                            |

> **Note**: Markdown supports rendering images via `![](img url)`,
> these images will be automatically downloaded when the message is viewed.
> Similarly, to why remote contents are blocked by default in e-mail clients,
> automatically downloading remote images could be used to collect information from the user.
>
> Also, if part of the message is interpolated from a malicious external source,
> the attacker could inject malformed markdown which leads to information disclosure.
>
> It is recommended to use `text/plain` to reduce possible security issues
> when using text from external sources like f.ex. output from scripts.

##### Example

```json
{
  "extras": {
    "client::display": {
      "contentType": "text/plain"
    }
  }
}
```

##### Support

| Client           | since  | description                                                     |
| ---------------- | ------ | --------------------------------------------------------------- |
| gotify/server ui | v2.0.5 | Uses [GitHub Flavored Markdown](https://github.github.com/gfm/) |
| gotify/android   | v2.0.7 | Uses [commonmark-spec](https://spec.commonmark.org/0.28/)       |

Author: Much better :+1: .
